lib/dal/src/component.rs

```diff
@@ -1084,17 +1090,98 @@ impl Component {
         for component in components {
             components_map.insert(component.id, HashSet::new());
 
-            for incomming_connection in component.incoming_connections(ctx).await? {
+            for incoming_connection in component.incoming_connections(ctx).await? {
                 components_map
                     .entry(component.id)
                     .or_default()
-                    .insert(incomming_connection.from_component_id);
+                    .insert(incoming_connection.from_component_id);
             }
         }
 
         debug!("build graph took {:?}", total_start.elapsed());
         Ok(components_map)
     }
+
+    pub async fn list_av_controlling_func_ids_for_id(
+        ctx: &DalContext,
+        component_id: ComponentId,
+    ) -> ComponentResult<HashMap<AttributeValueId, (FuncId, AttributeValueId, bool)>> {
+        let root_av_id: AttributeValueId =
+            Component::root_attribute_value_id(ctx, component_id).await?;
+
+        let mut av_queue = VecDeque::from([(root_av_id, None)]);
+        let mut result: HashMap<AttributeValueId, (FuncId, AttributeValueId, bool)> =
+            HashMap::new();
+
+        while let Some((av_id, maybe_parent_av_id)) = av_queue.pop_front() {
+            let prototype_id = AttributeValue::prototype_id(ctx, av_id).await?;
+            let func_id = AttributePrototype::func_id(ctx, prototype_id).await?;
+            let func = Func::get_by_id(ctx, func_id).await?;
+
+            let this_tuple = (func_id, av_id, func.is_dynamic());
+
+            // if av has a parent and parent is controlled by dynamic func (tuple.2), that's the controller
+            // else av controls itself
+            let controlling_tuple = if let Some(parent_av_id) = maybe_parent_av_id {
+                // We can unwrap because if we're on the child, we've added the parent to result
+                let parent_tuple = result.get(&parent_av_id).unwrap().clone();
```

Comment:
I don't know that we want an explicit panic.

Reply:
This should be fine. If we manage to not find something with the `get`, then on a previous iteration we somehow failed to insert into `result` at `1172`, while still finding results with the `av_queue.extend` at `1174`, and if that manages to happen, a `panic` is probably the right thing to do as we can't trust anything that's happening.

https://github.com/systeminit/si/pull/3454/files#diff-27851497f20cc332daa611bfe78589d554fbe07377859df57ce7dd53f0e999ddR1172-R1180

Reply:
Gotcha!